### PR TITLE
Align UX of packaging with the hydration UX

### PR DESCRIPTION
**Pull request by mortent**
This changes the output during `kpt pkg get` and `kpt pkg update` to be more aligned with the output from running the hydration pipeline. 

Fetching a composite package:
```
$ kpt pkg get git@github.com:mortent/kpt-composite-packages.git/kafka
Package "kafka":

Fetching git@github.com:mortent/kpt-composite-packages@main.
From github.com:mortent/kpt-composite-packages
 * branch            main       -> FETCH_HEAD
Adding package "kafka".

Package "kafka/zookeeper2":

Fetching https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Adding package "zookeeper".

Package "kafka/zookeeper2/zookeeper-networkpolicy":

Fetching https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Adding package "zookeeper-networkpolicy".

Package "kafka/zookeeper1":

Fetching https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Adding package "zookeeper".

Package "kafka/zookeeper1/zookeeper-networkpolicy":

Fetching https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Adding package "zookeeper-networkpolicy".

Fetched 5 package(s).
```

Updating same package:
```
$ kpt pkg update kafka
Package "kafka":

Fetching upstream from git@github.com:mortent/kpt-composite-packages@main.
From github.com:mortent/kpt-composite-packages
 * branch            main       -> FETCH_HEAD
Fetching origin from git@github.com:mortent/kpt-composite-packages@main.
From github.com:mortent/kpt-composite-packages
 * branch            22c9011801e190ce6cbfd7ae929b2ade6327a13b -> FETCH_HEAD

Package "kafka/zookeeper2":

Fetching upstream from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Fetching origin from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
Updating package "zookeeper2" with strategy "resource-merge".

Package "kafka/zookeeper2/zookeeper-networkpolicy":

Fetching upstream from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Fetching origin from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
Updating package "zookeeper-networkpolicy" with strategy "resource-merge".

Package "kafka/zookeeper1":

Fetching upstream from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Fetching origin from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
Updating package "zookeeper1" with strategy "resource-merge".

Package "kafka/zookeeper1/zookeeper-networkpolicy":

Fetching upstream from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            master     -> FETCH_HEAD
Fetching origin from https://github.com/mortent/kpt-packages@master.
From https://github.com/mortent/kpt-packages
 * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
Updating package "zookeeper-networkpolicy" with strategy "resource-merge".

Updated 5 package(s).
```

### CODE REVIEW

**Comment by frankfarzan** [comment]
> This changes the output during `kpt pkg get` and `kpt pkg update` to be more aligned with the output from running the hydration pipeline.
> 
> Fetching a composite package:
> 
> ```
> $ kpt pkg get git@github.com:mortent/kpt-composite-packages.git/kafka
> Fetching package "kafka" from "git@github.com:mortent/kpt-composite-packages" into "/Users/<user>/kafka"
> 
> Package "kafka": 
> Fetching git@github.com:mortent/kpt-composite-packages@main
> From github.com:mortent/kpt-composite-packages
>  * branch            main       -> FETCH_HEAD
> Adding package "kafka"
> 
> Package "kafka/zookeeper2": 
> Fetching https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Adding package "zookeeper"
> 
> Package "kafka/zookeeper2/zookeeper-networkpolicy": 
> Fetching https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Adding package "zookeeper-networkpolicy"
> 
> Package "kafka/zookeeper1": 
> Fetching https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Adding package "zookeeper"
> 
> Package "kafka/zookeeper1/zookeeper-networkpolicy": 
> Fetching https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Adding package "zookeeper-networkpolicy"
> ```
> 
> Updating same package:
> 
> ```
> $ kpt pkg update kafka
> Updating package in "/Users/<user>/kafka"
> 
> Package "kafka": 
> Fetching upstream from git@github.com:mortent/kpt-composite-packages@main
> From github.com:mortent/kpt-composite-packages
>  * branch            main       -> FETCH_HEAD
> Fetching origin from git@github.com:mortent/kpt-composite-packages@main
> From github.com:mortent/kpt-composite-packages
>  * branch            22c9011801e190ce6cbfd7ae929b2ade6327a13b -> FETCH_HEAD
> 
> Package "kafka/zookeeper2": 
> Fetching upstream from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Fetching origin from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
> Updating package "zookeeper2" with strategy "resource-merge"
> 
> Package "kafka/zookeeper2/zookeeper-networkpolicy": 
> Fetching upstream from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Fetching origin from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
> Updating package "zookeeper-networkpolicy" with strategy "resource-merge"
> 
> Package "kafka/zookeeper1": 
> Fetching upstream from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Fetching origin from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
> Updating package "zookeeper1" with strategy "resource-merge"
> 
> Package "kafka/zookeeper1/zookeeper-networkpolicy": 
> Fetching upstream from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            master     -> FETCH_HEAD
> Fetching origin from https://github.com/mortent/kpt-packages@master
> From https://github.com/mortent/kpt-packages
>  * branch            a0f76dcbe81d7ccd58511862db26a4f104c597d2 -> FETCH_HEAD
> Updating package "zookeeper-networkpolicy" with strategy "resource-merge"
> ```

Can you show the output of `render`? In `1.0.0-alpha.1` build there's a blank line after Package header:
```
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).
```

Has that changed? More important than whether to show blank line or not (presentation of package header will likely change as part #1850), is consistency.

**Comment by mortent** [comment]
> Can you show the output of `render`? In `1.0.0-alpha.1` build there's a blank line after Package header:
> 
> ```
> Package "wordpress/mysql": 
> 
> [RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
> [PASS] "gcr.io/kpt-fn/apply-setters:v0.1"
> 
> Package "wordpress": 
> 
> [RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
> [PASS] "gcr.io/kpt-fn/apply-setters:v0.1"
> 
> Successfully executed 2 function(s) in 2 package(s).
> ```
> 
> Has that changed? More important than whether to show blank line or not (presentation of package header will likely change as part #1850), is consistency by sharing the same presentation library.
> 
> In similar vein, do you want pkg commands to show the final success message with some useful statistics:
> 
> `Successfully executed 2 function(s) in 2 package(s).`

We can try to share more of the presentation logic, but then we need to have clearer rules for the structure we want to follow. In particular the use of blank lines that often should only be printed in certain situations are a bit awkward (like [this](https://github.com/GoogleContainerTools/kpt/blob/479c762b7c38238cbb2d0f64c6fa3f8347d8403a/internal/cmdrender/executor.go#L309) and [this](https://github.com/GoogleContainerTools/kpt/blob/479c762b7c38238cbb2d0f64c6fa3f8347d8403a/internal/cmdrender/executor.go#L274)) to handle as long as we don't have full separation between the presentation and the actual logic.

Based on the output from the `render` command , it seems like rules should be:
 * Package path is printed with a leading and ending newline, except when it is the first package printed.
 * Package path is the relative path from the users current working directory to each package (I find this a bit strange, this should be relative to the top-level package and not the current working directory).
 * Any output pertaining to a specific package is printed without any blank lines.
 * After an operation (render/get/update) is completed on a package, there is a blank line followed by a one-line summary/statistic.

Based on this, we could move some of the formatting for the package information into the printer on a somewhat higher level than `Printf`-style functions. The conditional newline before the package path is a little bit awkward, but not too bad.

Some other options:
 * I think we should avoid printing the full relative path from the users current working directory for every package path, so an alternative would be that include the path from the users working directory to the top-level package in the first line of output, and then only use package paths relative to the top-level package from there on out.

**Comment by phanimarupaka** [comment]
I don't think we need to print the path from user's working directory to root package. No matter from where the user runs the command, the output should be the same. 
```
➜  kpt git:(ConsistentPackageReferences) ✗ k fn render package-examples/wordpress/
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).
➜  kpt git:(ConsistentPackageReferences) ✗ cd package-examples 
➜  package-examples git:(ConsistentPackageReferences) ✗ k fn render wordpress/ 
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).   
➜  package-examples git:(ConsistentPackageReferences) ✗ cd wordpress 
➜  wordpress git:(ConsistentPackageReferences) ✗ k fn render
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).
➜  wordpress git:(ConsistentPackageReferences) ✗ k fn render .
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).
➜  wordpress git:(ConsistentPackageReferences) ✗ cd mysql 
➜  mysql git:(ConsistentPackageReferences) ✗ k fn render .
Package "mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 1 function(s) in 1 package(s).
➜  mysql git:(ConsistentPackageReferences) ✗ k fn render ../
Package "wordpress/mysql": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Package "wordpress": 

[RUNNING] "gcr.io/kpt-fn/apply-setters:v0.1"
[PASS] "gcr.io/kpt-fn/apply-setters:v0.1"

Successfully executed 2 function(s) in 2 package(s).
```

This is more related to how we handle `DisplayPath`. I am working on this and will send PR soon.

**Comment by mortent** [comment]
> I don't think we need to print the path from user's working directory to root package. No matter from where the user runs the command, the output should be the same.

I agree on this. My comment was just that in case we want to display the full relative path from the current directory, we should do that in one place and then keep the rest of the output clean (like in you example).

**Comment by phanimarupaka** [comment]
> > I don't think we need to print the path from user's working directory to root package. No matter from where the user runs the command, the output should be the same.
> 
> I agree on this. My comment was just that in case we want to display the full relative path from the current directory, we should do that in one place and then keep the rest of the output clean (like in you example).

I don't think we need to do that even once too. It has less usability as user already entered it as input so they are already aware of that path.

**Comment by phanimarupaka** [comment]
@mortent I tried this branch but the output is not as expected. It doesn't print info regarding subpackage. Can you please verify

```
➜  QuickStartTest k pkg get https://github.com/GoogleContainerTools/kpt/package-examples/wordpress@v0.1
Fetching package "package-examples/wordpress" from "https://github.com/GoogleContainerTools/kpt" into "/Users/pmarupaka/Testspace/QuickStartTest/wordpress"

Package "wordpress": 
Fetching https://github.com/GoogleContainerTools/kpt@v0.1
From https://github.com/GoogleContainerTools/kpt
 * tag               package-examples/wordpress/v0.1 -> FETCH_HEAD
Adding package "package-examples/wordpress"

Package "wordpress/mysql":
```

File: internal/cmdget/cmdget.go [hunk 1]
```diff
@@ -93,8 +93,9 @@ func (r *Runner) preRunE(_ *cobra.Command, args []string) error {
 
 func (r *Runner) runE(c *cobra.Command, _ []string) error {
 	const op errors.Op = "cmdget.runE"
-	fmt.Fprintf(c.OutOrStdout(), "fetching package %s from %s to %s\n",
-		r.Get.Git.Directory, r.Get.Git.Repo, r.Get.Destination)
+	pr := printer.FromContextOrDie(r.ctx)
+	pr.Printf("Fetching package %q from %q into %q\n",
+		strings.TrimPrefix(r.Get.Git.Directory, "/"), r.Get.Git.Repo, r.Get.Destination)
```
Comment (phanimarupaka) [inline internal/cmdget/cmdget.go:98]: `r.Get.Destination` this should be display path and not absolute path.

**Comment by mortent** [comment]
Updated the PR based on the feedback. The PR description has been updated with the UX after the changes.
This means:
 * Packaging now follows the same high-level structure as render, both regarding the use of blank lines and a summary at the end.
 * The printer package has a new `PrintPackage` function for printing the package path. It's a step towards having a higher-level printing api shared between the different components and reduce the use of `Printf`-style statements.

@phanimarupaka It has also been updated to fix the issue where the subpackage was printed incorrectly. 

This PR does not print any information about dependent packages. Doing this requires some larger changes in the update code, so I'll handle that in a separate PR.

**Review by frankfarzan — COMMENTED** [review]
UX lgtm, I'll let others review the code.

**Comment by phanimarupaka** [comment]
@mortent I tried it and this is how the output looks like.
```
Package "wordpress":

Fetching https://github.com/GoogleContainerTools/kpt@v0.1.
From https://github.com/GoogleContainerTools/kpt
 * tag               package-examples/wordpress/v0.1 -> FETCH_HEAD
Adding package "package-examples/wordpress".

Fetched 2 package(s).
```

It says fetch 2 packages but there is not reference to second package. Should it be 1 instead, till you make further changes to print info regarding subpackages ?

And regarding the empty line, you need not print it as I am removing it as part of this PR.

https://github.com/GoogleContainerTools/kpt/pull/1906/files#diff-50c35ee8085bce3d3e2879476b878032e9132d2ed0a25de904d4306fdc59870eR274

**Comment by phanimarupaka** [comment]
And do we have an issue to track the same for live commands ?

**Comment by mortent** [comment]
> And do we have an issue to track the same for live commands ?

I created an issue to track this: https://github.com/GoogleContainerTools/kpt/issues/1946.

**Comment by phanimarupaka** [comment]
```
Package "wordpress":
Fetching https://github.com/GoogleContainerTools/kpt@v0.1.
From https://github.com/GoogleContainerTools/kpt
 * tag               package-examples/wordpress/v0.1 -> FETCH_HEAD
Adding package "package-examples/wordpress".

Fetched 1 package(s).
```

Works fine. Thanks.

**Review by phanimarupaka — APPROVED** [review]

**Pull request merged**